core/feed_protocol/types.py

```diff
@@ -0,0 +1,160 @@
+from typing import Any, Dict, Generator, List, Optional, Tuple
+
+from pydantic import BaseModel
+
+from core.model import LicensePool, Work
+from core.model.edition import Edition
+from core.model.identifier import Identifier
+
+
+class FeedEntryType(BaseModel):
+    text: Optional[str] = None
+
+    class Config:
+        extra = "allow"
+        arbitrary_types_allowed = True
+
+    def add_attributes(self, attrs: Dict[str, Any]) -> None:
+        for name, data in attrs.items():
+            setattr(self, name, data)
+
+    def children(self) -> Generator[Tuple[str, "FeedEntryType"], None, None]:
+        """Yield all FeedEntryType attributes"""
+        for name, value in self:
+            if isinstance(value, self.__class__):
+                yield name, value
+        return
+
+
+class Link(FeedEntryType):
+    href: str
+    rel: Optional[str]
+    type: Optional[str]
+
+    # Additional types
+    role: Optional[str] = None
+    title: Optional[str] = None
+
+    def dict(self, **kwargs: Any) -> Dict[str, Any]:
+        kwargs["exclude_none"] = True
+        return super().dict(**kwargs)
+
+    def link_attribs(self) -> Dict[str, Any]:
+        d = dict(href=self.href)
+        for key in ["rel", "type"]:
+            if (value := getattr(self, key, None)) is not None:
+                d[key] = value
+        return d
+
+
+class IndirectAcquisition(BaseModel):
+    type: str
+    children: List["IndirectAcquisition"] = []
+
+
+class Acquisition(Link):
```

Comment:
If we are mapping our datamodel on to OPDS 2.x, then these additional fields in a acquisition link seem like that would be extensions that go into the [properties object ](https://readium.org/webpub-manifest/properties.html), rather then directly on the acquisition link itself.

Reply:
Under #1340 the serializer manages the output format so it works out fine, the models should just be used to capture the relevant data.